include/nbl/builtin/hlsl/tgmath.hlsl

```diff
@@ -0,0 +1,233 @@
+NBL_ALIAS_UNARY_FUNCTION(asin)
+NBL_ALIAS_UNARY_FUNCTION(atan)
+NBL_ALIAS_BINARY_FUNCTION(atan2)
+
+// Hyperbolic functions
+NBL_ALIAS_UNARY_FUNCTION(cosh)
+NBL_ALIAS_UNARY_FUNCTION(sinh)
+NBL_ALIAS_UNARY_FUNCTION(tanh)
+
+template<class T>
+T acosh(T x)
+{
+    return INTRINSIC_NAMESPACE(log)(x + INTRINSIC_NAMESPACE(sqrt)(x*x - T(1)));
+}
+
+// TODO numerically better impl of asinh
+//template<class T>
+//T asinh(T x)
+//{
+//    return INTRINSIC_NAMESPACE(log)(x + INTRINSIC_NAMESPACE(sqrt)(x*x + T(1)));
+//}
+
+template<class T>
+T atanh(T x)
+{
+    return T(0.5) * INTRINSIC_NAMESPACE(log)((T(1)+x)/(T(1)-x));
+}
+
+
+// Exponential and logarithmic functions
+NBL_ALIAS_UNARY_FUNCTION(exp)
+NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(frexp, int32_t)
+NBL_ALIAS_BINARY_FUNCTION(ldexp)
+NBL_ALIAS_UNARY_FUNCTION(log)
+NBL_ALIAS_UNARY_FUNCTION(log10)
+NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(modf, T)
+NBL_ALIAS_UNARY_FUNCTION(exp2)
+NBL_ALIAS_UNARY_FUNCTION(log2)
+NBL_ALIAS_UNARY_FUNCTION2(logb,log)
+
+//template<class T> 
+//T expm1(T x) 
+//{ 
+    // for better implementation, missing function for evaluate_polynomial
+    // https://www.boost.org/doc/libs/1_83_0/boost/math/special_functions/expm1.hpp
+    //or use std::expm1f, std::expm1l depending on size of T
+//}
+
+template<class T> 
+T log1p(T x) 
+{ 
+    // https://www.boost.org/doc/libs/1_83_0/boost/math/special_functions/log1p.hpp
+    /*if (x < T(-1))
+         "log1p(x) requires x > -1, but got x = %1%.
+        */
+    if (x == T(-1))
+    {
+        using uint_type = typename nbl::hlsl::numeric_limits<T>::uint_type;
+        return  -nbl::hlsl::numeric_limits<T>::infinity;
+    }
+    T u = T(1) + x;
+    if (u == T(1))
+        return x;
+    else
+        return INTRINSIC_NAMESPACE(log)(u) * (x / (u - T(1)));
+}
+
+template<class T>
+int32_t ilogb(T x) 
+{ 
+    using uint_type = typename nbl::hlsl::numeric_limits<T>::uint_type;
+    const int32_t shift = (impl::num_base<T>::float_digits-1);
+    const uint_type mask = ~(((uint_type(1) << shift) - 1) | (uint_type(1)<<(sizeof(T)*8-1)));
+    int32_t bits = (INTRINSIC_NAMESPACE(bit_cast)<uint_type, T>(x) & mask) >> shift; 
+    return bits + impl::num_base<T>::min_exponent - 2;
+}
+
+template<class T> 
+T scalbn(T x, int32_t n) 
+{ 
+    return x * INTRINSIC_NAMESPACE(exp2)(n); 
+}
+
+// Power functions
+NBL_ALIAS_BINARY_FUNCTION(pow)
+NBL_ALIAS_UNARY_FUNCTION(sqrt)
+// NBL_ALIAS_UNARY_FUNCTION(cbrt) // TODO
+
+template<class T>
+T hypot(T x, T y)
+{
+    return INTRINSIC_NAMESPACE(sqrt)(x*x+y*y);
+}
+
+// Floating-point manipulation functions 
+
+template<class T>
+T copysign(T x, T sign_)
+{
+    if ((x < 0 && sign_ > 0) || (x > 0 && sign_ < 0))
+        return -x;
+    return x;
+}
+
+// Generate quiet NaN (function)
+template<class T>
+T nan()
+{
+    using uint_type = typename nbl::hlsl::numeric_limits<T>::uint_type;
+    return INTRINSIC_NAMESPACE(bit_cast)<uint_type, T>(nbl::hlsl::numeric_limits<T>::quiet_NaN);
+}
```

Comment:
@keptsecret @Przemog1's ieee stuff does it better

Reply:
here: [ieee754.hlsl](https://github.com/Devsh-Graphics-Programming/Nabla/blob/cpp_intrinsics_improvements/include/nbl/builtin/hlsl/ieee754.hlsl)

Reply:
@devshgraphicsprogramming maybe @keptsecret should merge [this pr?](https://github.com/Devsh-Graphics-Programming/Nabla/pull/804)

Reply:
probably yes